In [53]:
import altair as alt
import pandas as pd
import os

alt.data_transformers.enable("default", max_rows=None)
from altair_data_server import data_server
alt.data_transformers.enable('data_server')
from vega_datasets import data
alt.renderers.enable('default')
alt.renderers.enable('png')

RendererRegistry.enable('png')

In [2]:
import altair as alt

In [3]:
import graph_maker

In [4]:
import graph_maker_1

# Who is the best Power Lifter ( Model Lifter)?

## Introduction

Powerlifting is a strength sport that consists of three attempts at maximal weight on three lifts: squat, bench press, and deadlift. As in the sport of Olympic weightlifting, it involves the athlete attempting a maximal weight single lift of a barbell loaded with weight plates. Powerlifting evolved from a sport known as "odd lifts", which followed the same three-attempt format but used a wider variety of events, akin to strongman competition. Eventually odd lifts became standardized to the current three.

This analysis is based on data set propvided by owerlifting.org The OpenPowerlifting project aims to create a permanent, accurate, convenient, accessible, open archive of the world's powerlifting data.

### Question(s) of interests



1. How is the  performance of Power Lifters impacted by different Equipment? Are certain equipment better than others for certain type of lifts?
2. How is the  performance of Power Lifters impacted by their Gender? How powerlifters perform when measured against the total weight lifted and the weight lifted as compared to their Body Weight? 
3. What can be a good Metric to measure power lifters performance?
4. How the power lifter performance changes with age?

### Data Wrangling

In [5]:
url_pwr_lft ='https://raw.githubusercontent.com/UBC-MDS/exploratory-data-viz/main/data/power_lifting.csv'

In [6]:
pwr_lft_df=pd.read_csv(url_pwr_lft)

In [7]:
pwr_lft_df.describe()

MeetID            Age   BodyweightKg     Squat4Kg  \
count  386414.000000  147147.000000  384012.000000  1243.000000   
mean     5143.015804      31.668237      86.934912   107.036404   
std      2552.099838      12.900342      23.140843   166.976620   
min         0.000000       5.000000      15.880000  -440.500000   
25%      2979.000000      22.000000      70.300000    87.500000   
50%      5960.000000      28.000000      83.200000   145.000000   
75%      7175.000000      39.000000     100.000000   212.500000   
max      8481.000000      95.000000     242.400000   450.000000   

         BestSquatKg     Bench4Kg    BestBenchKg  Deadlift4Kg  BestDeadliftKg  \
count  298071.000000  1962.000000  356364.000000  2800.000000   317847.000000   
mean      176.569941    45.722905     118.347509   113.597193      195.040633   
std        69.222785   151.668221      54.848850   170.201657       61.580675   
min      -477.500000  -360.000000    -522.500000  -461.000000     -410.000000   
25%       127.500000   -90.000000      79.380000   110.000000      147.500000   
50%       174.630000    90.250000     115.000000   157.500000      195.000000   
75%       217.720000   167.500000     150.000000   219.990000      238.140000   
max       573.790000   378.750000     488.500000   418.000000      460.400000   

             TotalKg          Wilks  
count  363237.000000  362194.000000  
mean      424.000249     301.080601  
std       196.355147     116.360396  
min        11.000000      13.730000  
25%       272.160000     237.380000  
50%       424.110000     319.660000  
75%       565.000000     379.290000  
max      1365.310000     779.380000

In [8]:
pwr_lft_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 386414 entries, 0 to 386413
Data columns (total 17 columns):
MeetID            386414 non-null int64
Name              386414 non-null object
Sex               386414 non-null object
Equipment         386414 non-null object
Age               147147 non-null float64
Division          370571 non-null object
BodyweightKg      384012 non-null float64
WeightClassKg     382602 non-null object
Squat4Kg          1243 non-null float64
BestSquatKg       298071 non-null float64
Bench4Kg          1962 non-null float64
BestBenchKg       356364 non-null float64
Deadlift4Kg       2800 non-null float64
BestDeadliftKg    317847 non-null float64
TotalKg           363237 non-null float64
Place             385322 non-null object
Wilks             362194 non-null float64
dtypes: float64(10), int64(1), object(6)
memory usage: 50.1+ MB


In [9]:
pwr_lft_df.head(5)

MeetID              Name Sex   Equipment   Age     Division  BodyweightKg  \
0       0  Angie Belk Terry   F       Wraps  47.0    Mst 45-49         59.60   
1       0       Dawn Bogart   F  Single-ply  42.0    Mst 40-44         58.51   
2       0       Dawn Bogart   F  Single-ply  42.0  Open Senior         58.51   
3       0       Dawn Bogart   F         Raw  42.0  Open Senior         58.51   
4       0      Destiny Dula   F         Raw  18.0   Teen 18-19         63.68   

  WeightClassKg  Squat4Kg  BestSquatKg  Bench4Kg  BestBenchKg  Deadlift4Kg  \
0            60       NaN        47.63       NaN        20.41          NaN   
1            60       NaN       142.88       NaN        95.25          NaN   
2            60       NaN       142.88       NaN        95.25          NaN   
3            60       NaN          NaN       NaN        95.25          NaN   
4          67.5       NaN          NaN       NaN        31.75          NaN   

   BestDeadliftKg  TotalKg Place   Wilks  
0           70.31   138.35     1  155.05  
1          163.29   401.42     1  456.38  
2          163.29   401.42     1  456.38  
3             NaN    95.25     1  108.29  
4           90.72   122.47     1  130.47

In [10]:
pwr_lft_df.columns

Index(['MeetID', 'Name', 'Sex', 'Equipment', 'Age', 'Division', 'BodyweightKg',
       'WeightClassKg', 'Squat4Kg', 'BestSquatKg', 'Bench4Kg', 'BestBenchKg',
       'Deadlift4Kg', 'BestDeadliftKg', 'TotalKg', 'Place', 'Wilks'],
      dtype='object')

My columns of interest are 'Age',BodyweightKg','BestSquatKg','BestBenchKg',BestDeadliftKg', 'TotalKg','Wilks.

In [11]:
pwr_lft_df[['BestDeadliftKg','BestBenchKg','BestSquatKg']].describe()

BestDeadliftKg    BestBenchKg    BestSquatKg
count   317847.000000  356364.000000  298071.000000
mean       195.040633     118.347509     176.569941
std         61.580675      54.848850      69.222785
min       -410.000000    -522.500000    -477.500000
25%        147.500000      79.380000     127.500000
50%        195.000000     115.000000     174.630000
75%        238.140000     150.000000     217.720000
max        460.400000     488.500000     573.790000

In [12]:
pwr_lft_df[['BestDeadliftKg','BestBenchKg','BestSquatKg']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 386414 entries, 0 to 386413
Data columns (total 3 columns):
BestDeadliftKg    317847 non-null float64
BestBenchKg       356364 non-null float64
BestSquatKg       298071 non-null float64
dtypes: float64(3)
memory usage: 8.8 MB


In [13]:
# As We can see above the minu=imums valus are negative values and null valus , however, the functions that I have cdedconsiderv only positive values and drop Nan Values

### How is the  performance of Power Lifters impacted by different Equipment? Are certain equipment better than others for certain type of lifts?

Since it’s inception, the goal of powerlifting has always been to lift the most weight in three lifts — squat, deadlift, and bench press. The Lifts can be broadly classified into Raw and Equipped Type. Raw powerlifting means lifting with little to no additional equipment."Equipped” powerlifting means using additional gear to both protect the body and aid in lifting more weight. Power Lifting Gear include Straps,wraps, single-ply and Muti-ply Fabrics.

How has the Equipment effected the Performance of the Power Lifters in different Lift Categories?

In [14]:
lifts_distribution= alt.Chart(url_pwr_lft,title ="Distribution of Lifts").mark_bar( ).encode(alt.X('Equipment:N'),alt.Y('count()'))
lifts_distribution= lifts_distribution.save('lifts_distribution.png')
lifts_distribution

alt.Chart(...)

In [15]:
pwr_lft_df[['BestSquatKg','BestDeadliftKg','BestBenchKg','Equipment']].groupby(['Equipment']).count()

BestSquatKg  BestDeadliftKg  BestBenchKg
Equipment                                           
Multi-ply          5590            6408         7842
Raw              136210          155816       172729
Single-ply       110894          110995       130962
Straps                0               7            0
Wraps             45377           44621        44831

Mean of Weight Lifted for Different Lift Types

In [16]:
# The Sample Size of Starps is too small to draw meaningful conclusions so we will not consider it
pwr_lft_df=pwr_lft_df[pwr_lft_df['Equipment']!='Straps']

In [17]:
mean_df =pwr_lft_df[['BestSquatKg','BestDeadliftKg','BestBenchKg','Equipment']].groupby(['Equipment']).mean().reset_index()
mean_df.head(1)

Equipment  BestSquatKg  BestDeadliftKg  BestBenchKg
0  Multi-ply    294.71266      249.523414   202.032113

In [18]:
table_mean =pd.melt(mean_df, id_vars=['Equipment'], value_vars=['BestSquatKg', 'BestDeadliftKg', 'BestBenchKg'])
table_mean.head(1)

Equipment     variable      value
0  Multi-ply  BestSquatKg  294.71266

In [19]:
# Let us plot the mean of Weight Lifted for different Lift Types using different 'equipment'
#to get a basic idea of how the equipment has impacted lifts performance
chart_mean_lifts = alt.Chart(table_mean,title ='Mean Kilograms of different Lift Types').mark_bar(
    opacity=0.3,
    ).encode(
    column = alt.Column('variable'),
    x = alt.X('Equipment:N'),
    y = alt.Y('value:Q'),
    color = alt.Color('variable:N'),
).properties(width= 150, height = 150)
chart_mean_lifts

alt.Chart(...)

As we can see generally powerlifting performance improves as the powerlifter gets “more equipped,”.The average weight Lift for equipped(Multi-ply, single-ply, Wraps) exceeds the Raw type of Lift for Bench and Squat Type of the Lifts with Multi-ply seeming to have much better average performance than other Equipment Type. 

For the Deadlift Category, however, the mean value of the Raw is slightly higher(almost same) as that of Single-ply for Dead Lift. Hence all equipped Lifts may not always be better than Raw Lifts?The means of the Wraps and Multi-ply are also very close, hence, further analysis is required to uderstand the impact of different Equipment types on Lifts.

In [20]:
mean_df[(mean_df['Equipment']=='Multi-ply') | (mean_df['Equipment']=='Wraps')][['BestDeadliftKg']]

BestDeadliftKg
0      249.523414
3      213.524099

In [21]:
mean_df[(mean_df['Equipment']=='Multi-ply') | (mean_df['Equipment']=='Wraps')][['BestDeadliftKg']].describe()[2:3]

BestDeadliftKg
std        25.45536

In [22]:
mean_df[(mean_df['Equipment']=='Raw') | (mean_df['Equipment']=='Single-ply')][['Equipment','BestDeadliftKg']]

Equipment  BestDeadliftKg
1         Raw      190.679162
2  Single-ply      190.575193

To further understand the impact of different equipment on different lifts and athlete performance, let us plot different Lifts against athlete weights. We will approximate our data using a linear regressor as it will approximate how the performance of the athletes with different weights /in different weight categories are effected by different Equipment.


In [23]:
pwr_lft_df[['BodyweightKg','BestSquatKg','BestDeadliftKg','BestBenchKg']].describe()

BodyweightKg    BestSquatKg  BestDeadliftKg    BestBenchKg
count  384005.000000  298071.000000   317840.000000  356364.000000
mean       86.934371     176.569941      195.036379     118.347509
std        23.140396      69.222785       61.574514      54.848850
min        15.880000    -477.500000     -410.000000    -522.500000
25%        70.300000     127.500000      147.500000      79.380000
50%        83.200000     174.630000      195.000000     115.000000
75%       100.000000     217.720000      238.140000     150.000000
max       242.400000     573.790000      460.400000     488.500000

In [24]:
pwr_lft_df[['BodyweightKg','BestSquatKg','BestDeadliftKg','BestBenchKg']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 386407 entries, 0 to 386413
Data columns (total 4 columns):
BodyweightKg      384005 non-null float64
BestSquatKg       298071 non-null float64
BestDeadliftKg    317840 non-null float64
BestBenchKg       356364 non-null float64
dtypes: float64(4)
memory usage: 14.7 MB


In [25]:
#As we can see above the minimum of the different lifts is negative and there is missing data as well. 
#The following code does not consider negative values or Nan Values while creating Plots

In [26]:
best_squat_body_weight = graph_maker.functionChart (url_pwr_lft,'BodyweightKg','BestSquatKg',600)

In [27]:
best_daedlift_body_weight = graph_maker.functionChart (url_pwr_lft,'BodyweightKg','BestDeadliftKg',450)

In [28]:
best_bench_body_weight = graph_maker.functionChart (url_pwr_lft,'BodyweightKg','BestBenchKg',600)

In [29]:
best_squat_body_weight | best_daedlift_body_weight  | best_bench_body_weight 

alt.HConcatChart(...)

These regressors confirm that 'Multi-ply' exceeds in performance when the Lift is Bench or Squat Type. However, for best dead lift Wraps exceed in performance than 'Multi-ply', specially in higher weight categories.

Also Raw Perform much better than Single-ply in many instances and as good as Multi-ply in few instances ( confirming that all Equipped Lifts are not better

The relationship between weight lifted and Bodyweight is a positive. I guess this would mean that the heavier
athletes are more stronger and muscular too!

**Straps not considered in analysis and sample size is too small

### How is the  performance of Power Lifters impacted by their Gender? How powerlifters perform when measured against the total weight lifted and the weight lifted as compared to their Body Weight? 

How is Power Lifter performance effected by sex? Is the Distribution for the Male and Female Lifters more like a Pareto distribution, where 'few at top' vastly outperforms the 'few at bottom'? Is it more like normal distribution wher the performance is 'evenly' distributed along an average 'mean' ?

In [30]:
numeric_cols=['BestBenchKg','BestSquatKg','BestDeadliftKg','TotalKg']
selector_1 = alt.selection_single(empty='all', fields=['Sex'])

color_scale_1 = alt.Scale(domain=['M', 'F'],
                        range=['#1FC3AA', '#8624F5'])

base_1 = alt.Chart(url_pwr_lft).properties(
    width=300,
    height=300
).add_selection(selector_1)

hists1 = base_1.mark_bar(clip= True, opacity=0.5, thickness=100).encode(
    x=alt.X(alt.repeat('repeat'),type='quantitative',
            bin=alt.Bin(step=5), # step keeps bin size the same
            scale=alt.Scale(domain=[10,800])),
    y=alt.Y('count()',
            stack=None,
            scale=alt.Scale(domain=[0,15000])),
    color=alt.Color('Sex:N',
                    scale=color_scale_1)
).transform_filter(
    selector_1
).properties(height=300, width =300).repeat(repeat =numeric_cols,columns =2)

hists1 = hists1.properties(title ={"text" :["Distributions of BestBench(kg), BestSquat(Kg), BestWeight(Kg), TotalKg for Male and Female"]})
hists1

alt.RepeatChart(...)

Based on above the distribution in various lift categories for different genders resemble Normal distributions. Men outperforms Women, the outliers in Women compete with the Average and bottom in the Lift Categories and not with the top.


But generally Men are heavier than Women . As noted above, there is a positive relationship between  te weight lifted and body weight, let us explore further.

In [31]:
pwr_lft_df[pwr_lft_df['Sex']== 'M'][['BodyweightKg']].describe()

BodyweightKg
count  296924.000000
mean       92.074109
std        22.216335
min        17.960000
25%        75.200000
50%        89.540000
75%       105.000000
max       242.400000

In [32]:
pwr_lft_df[pwr_lft_df['Sex']== 'F'][['BodyweightKg']].describe()

BodyweightKg
count  87081.000000
mean      69.409177
std       16.768542
min       15.880000
25%       58.100000
50%       66.500000
75%       76.600000
max      184.900000

In [33]:
bdywgt_bestsqt_male_female = graph_maker_1.functionChart1 (url_pwr_lft,'BodyweightKg','BestSquatKg')
#chart777.tooltip=[alt.Tooltip("'BodyweightKg:Q", title="Body Weight Kilograms"),alt.Tooltip("BestSquatKg:Q", title="Best Squat Kilogram")]
bdywgt_bestsqt_male_female


alt.FacetChart(...)

In [34]:
bdywgt_bestdeadlift_male_female = graph_maker_1.functionChart1(url_pwr_lft,'BodyweightKg','BestDeadliftKg')
bdywgt_bestdeadlift_male_female

alt.FacetChart(...)

In [35]:
dywgt_bestbenchliift_male_female =  graph_maker_1.functionChart1 (url_pwr_lft,'BodyweightKg','BestBenchKg')
dywgt_bestbenchliift_male_female

alt.FacetChart(...)

Although Men is heavier than Women, they also lift multiple times more weight of their Body Weight than Women , generally.

On closer look of the plots, we can see from the plots above that the  (absolute) maximum weight lifted ( by best performing athletes)increases with the increasing weight of the athletes, it then reaches a paek and then level off or may actually decrease slightly for ther athletes in higher end of body weight class. 

When comparing how much an atlete lift compared to his/her body weight,Lighter athletes lift many times more multiple weight of their body weight than heavier athletes... this lead us to a question of what it means to be a stronger powerlifter.. one who lifts more as compared to his/her body weight or one who lifs more (absolute) weight compared to the others?

Clearly, bodyweight multipliers aren’t a very good standard for comparing relative strength across a wide range of body weight... it will always make lighter athletes more stronger than heavier one.. because they can lift many time more wight than heavier ones.

The strength of an athlete is not a simple factor of how many multiples he/she can lift of its body weight, what can then be a good metric for predicting the strength of athletes? 

## What can be a good Metric to measure power lifters performance?

Simple calculations based on how many times bodyweight you can lift is not a good indication of the strength of powerlifters, as this does not scale in a linear fashion. It would highly disadvantage the heavier lifter versus the lighter lifter.

Robert Wilks of Powerlifting Australia has provided another metric for describing the strebgth of powerlifters. let us see how good it can predict the strength of a power lifting athlete?

In [36]:
wilks_dstrbn = alt.Chart(url_pwr_lft).mark_bar().encode(
     alt.X("Wilks:Q", bin=alt.Bin(maxbins=40),title='Wilks Score'),
     alt.Y('count()',title='Number of Athletes')).properties(title ='Wilks Distribution')
wilks_dstrbn

alt.Chart(...)

In [37]:
wilks_dstrbn.facet('Sex:N')

alt.FacetChart(...)

In [38]:
wilks_dstrbn_best = alt.Chart(url_pwr_lft).mark_bar().encode(
     alt.X("Wilks:Q", bin=alt.Bin(maxbins=40),title='Wilks Score'),
     alt.Y('count()',title='Number of Athletes')).properties(title ='Wilks Distribution for scores greater than 320').transform_filter(alt.datum.Wilks >320)
wilks_dstrbn_best

alt.Chart(...)

As we can see from above the number of athletes with higher wilkis score decrease "exponentially" as we move above Willkis score of 320. Are these the highest perfoemers, let us explore further by
looking at the Wilkis score of top performers. 

Let us look at the top lifts (tota kgs lifted as well as the kgs lifted against body weight) abd Wilkis score of the highest lifters

In [39]:
#Relative Strength
def squatBody(x):
    return x['BestSquatKg'] / x['BodyweightKg']

def benchBody(x):
    return x['BestBenchKg'] / x['BodyweightKg']

def deadliftBody(x):
    return x['BestDeadliftKg'] / x['BodyweightKg']

def totalLiftBody(x):
    return x['TotalKg'] / x['BodyweightKg']


In [40]:

pwr_lft_df['Squat/BW'] = pwr_lft_df.apply(squatBody,axis=1)
pwr_lft_df['Bench/BW'] = pwr_lft_df.apply(benchBody,axis=1)
pwr_lft_df['Deadlift/BW'] =pwr_lft_df.apply(deadliftBody,axis=1)
pwr_lft_df['Total/BW'] = pwr_lft_df.apply(totalLiftBody,axis=1)


It looks like that there is a sharp decrease in the number of athletes as Wilks score increases above 320. Are these the highest performing athletes?

In [41]:
pwr_lft_df.head(1)

MeetID              Name Sex Equipment   Age   Division  BodyweightKg  \
0       0  Angie Belk Terry   F     Wraps  47.0  Mst 45-49          59.6   

  WeightClassKg  Squat4Kg  BestSquatKg  ...  BestBenchKg  Deadlift4Kg  \
0            60       NaN        47.63  ...        20.41          NaN   

   BestDeadliftKg  TotalKg  Place   Wilks  Squat/BW  Bench/BW  Deadlift/BW  \
0           70.31   138.35      1  155.05  0.799161   0.34245     1.179698   

   Total/BW  
0  2.321309  

[1 rows x 21 columns]

In [42]:
pwr_lft_df[pwr_lft_df['Sex']== 'M'][['Name','BodyweightKg','TotalKg','Total/BW','Wilks']].sort_values(by='TotalKg', ascending = False)[:20]

Name  BodyweightKg  TotalKg   Total/BW   Wilks
57011         Dave Hoff        131.09  1365.31  10.415058  771.07
53198         Dave Hoff        122.97  1363.05  11.084411  779.38
162480  Donnie Thompson        174.69  1360.78   7.789685  735.88
55694         Dave Hoff        129.27  1344.90  10.403806  761.44
163054        Dave Hoff        126.55  1342.63  10.609482  763.22
164936        Dave Hoff        123.79  1342.63  10.846030  766.63
53735         Dave Hoff        126.19  1324.49  10.495998  753.33
164547        Dave Hoff        124.83  1322.22  10.592165  753.67
162475        Dave Hoff        125.00  1319.95  10.559600  752.17
164551       AJ Roberts        137.71  1295.01   9.403892  725.48
6299          Dave Hoff        123.30  1290.00  10.462287  737.19
162476       AJ Roberts        140.00  1283.67   9.169071  717.32
161154       AJ Roberts        138.16  1281.40   9.274754  717.49
156733        Dave Hoff        126.28  1277.50  10.116408  726.50
298158    Blaine Sumner        170.68  1272.50   7.455472  690.71
161151        Dave Hoff        120.29  1272.33  10.577188  731.08
61917     Blaine Sumner        167.51  1271.50   7.590592  692.21
53204     Jake Anderson        142.61  1270.06   8.905827  707.79
53205     Shane Hammock        143.74  1270.06   8.835815  706.98
161853       AJ Roberts        137.53  1260.99   9.168836  706.56

In [43]:
pwr_lft_df[pwr_lft_df['Sex']== 'M'][['Name','BodyweightKg','TotalKg','Total/BW','Wilks']].sort_values(by='Total/BW', ascending = False)[:20]

Name  BodyweightKg  TotalKg   Total/BW   Wilks
164482        Lones Green         63.19   864.09  13.674474  703.82
164412        Lones Green         63.19   864.09  13.674474  703.82
164447        Lones Green         63.19   864.09  13.674474  703.82
48136       Kalle Räsänen         82.50  1100.00  13.333333  736.90
63607   Sergey Fedosienko         55.86   742.50  13.292159  677.59
159715       Shawn Frankl         90.00  1192.95  13.255000  761.57
63858   Sergey Fedosienko         55.49   732.50  13.200577  672.85
57830   Sergey Fedosienko         54.92   720.00  13.109978  668.17
59755   Sergey Fedosienko         58.06   761.00  13.107131  669.14
58929   Sergey Fedosienko         57.63   755.00  13.100816  668.56
48231       Kalle Räsänen         89.80  1175.00  13.084633  750.96
59110   Sergey Fedosienko         56.82   743.00  13.076382  666.91
53260         Nick Graves         60.37   789.25  13.073546  669.38
62763   Sergey Fedosienko         58.49   762.50  13.036416  665.83
57657   Sergey Fedosienko         54.94   715.00  13.014197  663.29
59874   Sergey Fedosienko         57.60   745.00  12.934028  660.03
58263   Sergey Fedosienko         57.03   737.50  12.931790  659.63
58033   Sergey Fedosienko         57.46   730.00  12.704490  648.24
56051       Wes McCormick         74.39   943.47  12.682753  676.13
162477  Michael Cartinian         82.50  1043.26  12.645576  698.89

In [44]:
pwr_lft_df[pwr_lft_df['Sex']== 'F'][['Name','BodyweightKg','TotalKg','Total/BW','Wilks']].sort_values(by='TotalKg', ascending = False)[:20]

Name  BodyweightKg  TotalKg   Total/BW   Wilks
162483  Laura Phelps-Sweatt         75.00   816.47  10.886267  776.17
158866  Laura Phelps-Sweatt         82.50   802.87   9.731758  722.56
164532  Laura Phelps-Sweatt         74.75   802.86  10.740602  764.83
160104  Laura Phelps-Sweatt         75.00   796.05  10.614000  756.76
53186   Laura Phelps-Sweatt         82.53   789.25   9.563189  710.17
162405  Laura Phelps-Sweatt         73.94   787.50  10.650527  755.41
159000  Laura Phelps-Sweatt         74.84   780.18  10.424639  742.66
5458           April Mathis        125.00   772.50   6.180000  613.28
161392      Melissa Garrett        157.40   771.11   4.899047  592.70
63677          Bonica Lough        130.97   763.00   5.825762  600.68
159001      Melissa Garrett        162.84   759.77   4.665746  587.82
59102    Olga Gemaletdinova        120.95   753.00   6.225713  601.30
64035          Bonica Lough        132.95   750.00   5.641219  588.82
164776         April Mathis        119.66   748.43   6.254638  598.79
58027        Galina Karpova        128.26   747.50   5.828006  590.72
57824        Galina Karpova        132.06   745.00   5.641375  585.62
2522           April Mathis           NaN   730.28        NaN     NaN
172461     Shawna Mendelson         80.74   725.75   8.988729  660.71
156263     Samantha Coleman        160.57   725.75   4.519836  559.41
58255        Galina Karpova        128.35   721.00   5.617452  569.70

In [45]:
pwr_lft_df[pwr_lft_df['Sex']== 'F'][['Name','BodyweightKg','TotalKg','Total/BW','Wilks']].sort_values(by='Total/BW', ascending = False)[:20]

Name  BodyweightKg  TotalKg   Total/BW   Wilks
53119           Heidi Howar         59.24   678.12  11.446995  763.55
162483  Laura Phelps-Sweatt         75.00   816.47  10.886267  776.17
156240           Rheta West         65.68   709.87  10.808009  739.03
386161           Rheta West         66.04   712.14  10.783465  738.43
164532  Laura Phelps-Sweatt         74.75   802.86  10.740602  764.83
53718   Laura Phelps-Sweatt         66.27   709.87  10.711785  734.21
48142      Marianne Kosonen         47.63   510.00  10.707537  679.30
62739         Wei-Ling Chen         46.75   500.00  10.695187  675.03
162405  Laura Phelps-Sweatt         73.94   787.50  10.650527  755.41
160104  Laura Phelps-Sweatt         75.00   796.05  10.614000  756.76
59686         Wei-Ling Chen         45.60   483.50  10.603070  664.38
53719            Rheta West         65.50   691.73  10.560763  721.60
59998      Yukako Fukushima         46.09   484.00  10.501193  660.09
172787           Rheta West         67.22   705.34  10.493008  722.04
63648      Yukako Fukushima         46.44   485.00  10.443583  657.91
159000  Laura Phelps-Sweatt         74.84   780.18  10.424639  742.66
59035         Wei-Ling Chen         46.56   480.50  10.320017  650.60
59692     Natalia Salnikova         51.36   530.00  10.319315  667.05
63647     Natalia Salnikova         50.88   525.00  10.318396  665.54
60005     Natalia Salnikova         50.96   525.00  10.302198  664.74

As we can see from above a higther wilker scale athlete signifies a more capable Power Lifter. The top performers are the onec with the higher Wilks score 

In [46]:
slider = alt.binding_range(
    step=50,
    min=50, max=700)

select_year_slider = alt.selection_single(
    fields=['Wilks'],
    bind=slider,
    init={'Wilks' : 500},
    name='Select')


wills_plot_absolute_strength =alt.Chart(url_pwr_lft).mark_circle(size =10,clip=True).encode( 
                alt.X('Wilks:Q',impute=alt.ImputeParams(value=None),scale=alt.Scale(zero=False),axis =alt.Axis (title ='Wilks Score')),
                alt.Y('TotalKg:Q',impute=alt.ImputeParams(value=None),scale=alt.Scale(zero=False),axis =alt.Axis (title ='Totalkg(Kgs)')),
                [alt.Tooltip("Wilks:Q",title="Wilks Score"),alt.Tooltip('TotalKg:Q', title="Total Kilogram")],
                color=alt.condition(abs(alt.datum.Wilks - select_year_slider.Wilks) < 30, alt.value("navy"), alt.value("lightgray"))).properties(width=300,height=0,title="Scatter plot of Total Kilograms Lifted against Wilkis Score").add_selection(select_year_slider)
                

wills_plot_absolute_strength

alt.Chart(...)

In [47]:
pwr_lft_df[['Wilks','Total/BW']]

Wilks  Total/BW
0       155.05  2.321309
1       456.38  6.860708
2       456.38  6.860708
3       108.29  1.627927
4       130.47  1.923210
...        ...       ...
386409  202.60  3.059518
386410  185.77  2.709629
386411  211.17  3.055117
386412  181.85  2.525053
386413  205.18  2.800093

[386407 rows x 2 columns]

In [48]:
alt.data_transformers.disable_max_rows()
slider = alt.binding_range(
    step=50,
    min=50, max=700)

select_year_slider = alt.selection_single(
    fields=['Wilks'],
    bind=slider,
    init={'Wilks' : 500},
    name='Select')


wills_plot_rel_strength =alt.Chart(pwr_lft_df[['Wilks','Total/BW']]).mark_circle(size =10,clip=True).encode( 
                alt.X('Wilks',impute=alt.ImputeParams(value=None),scale=alt.Scale(zero=False),axis =alt.Axis (title ='Wilks Score')),
                alt.Y('Total/BW',impute=alt.ImputeParams(value=None),scale=alt.Scale(zero=False),axis =alt.Axis (title ='Totalkg Relative to Body Weight(Kgs)')),
                [alt.Tooltip("Wilks:Q",title="Wilks Score"),alt.Tooltip('Total/BW', title="Total Kilogram muliples of Body Weight")],
                color=alt.condition(abs(alt.datum.Wilks - select_year_slider.Wilks) < 30, alt.value("navy"), alt.value("lightgray"))).properties(width=300,height=0,title="Scatter plot of Total Kilograms Lifted against Wilkis Score").add_selection(select_year_slider)

  
wills_plot_rel_strength              



alt.Chart(...)

As we can see from above a higther wiolker scole athlete signifies a more cpable Power Lifter.

The wilker scale does not seem to discriminate between the light and heavy lifters. Athletes with Wilks score in the highest band performs better on the absolute strength(Totalkgs) and relative strength(Total Kg/BW).re. So Wilkis score is a better measure of the aythlete strength and powerlifting ability.  

Let us ginally look at the Performance of different athletes against Age.

## How the power lifter performance changes with age?

In [49]:
age_dstrbn = alt.Chart(url_pwr_lft).mark_bar().encode(
     alt.X("Age:Q", bin=alt.Bin(maxbins=40),title='Age'),
     alt.Y('count()',title='Number of Athletes')).properties(title ='Wilks Distribution')
age_dstrbn

alt.Chart(...)

In [50]:
age_dstrbn.facet('Sex:N')

alt.FacetChart(...)

In [51]:
chart777 =  graph_maker_1.functionChart1 (url_pwr_lft,'Age','TotalKg')
chart777

alt.FacetChart(...)

As we can ee from the above the peak performance for Men is between the age of 27, wheraes for Women , itv is between 27-32. Also the ability to lift weight decreases with age.

## DashBoard:

In [52]:
(wills_plot_absolute_strength | wills_plot_rel_strength) & (best_squat_body_weight | best_daedlift_body_weight  | best_bench_body_weight ).properties(title ="How Wilks and equipment selection effect powerllifting performance ") 

alt.VConcatChart(...)

# Concluding Remarks

For Males , Based on the assesment above a Modell Lifter would be around 120 KG Male with high Willkers Score, around 27 years of age, who is familiar with the advances in fabrics technology so as to utilize the new technology to increase the weight that he can lift.

For FeMales , Based on the assesment above a Modell Lifter would be around 80 KG Male with high Willkers Score, around 27 years of age, who is familiar with the advances in  fabrics technology so as to utilize the new technology to increase the weight that he can lift.

## Helpful Resources and Documentation

Without mentioning any specific name, I would like to thank the online resources that help me during this endavaour